Replace local web server setup for web editor with a Python-based solution

Calinou opened the pull request:
Follow-up to https://github.com/godotengine/godot/pull/56977.

This makes it easier to set up, as you always have Python installed when building Godot. On the other hand, you don't always have Node.js + npm installed (and you may not want to spend time running `npm install`).

I tested this locally and made sure it works.

Future improvements:

- [x] Add a `serve=yes` SCons option that automatically extracts the generated ZIP and calls the serve script after building. (This would only work when using `platform=javascript.`)
  - Done in https://github.com/godotengine/godot/pull/67325
- [x] Document building and usage instructions for all platforms (I'll do this in separate PRs, both for the main repository and godot-docs).
  - Done in https://github.com/godotengine/godot/pull/64889 and https://github.com/godotengine/godot/pull/64886.

Calinou committed:
Replace local web server setup for web editor with a Python-based solution

This makes it easier to set up, as you always have Python installed
when building Godot. On the other hand, you don't always have Node.js
+ npm installed (and you may not want to spend time running `npm install`).

Co-authored-by: Fabio Alessandrelli <fabio.alessandrelli@gmail.com>

yedpodtrzitko commented on platform/javascript/serve.py:
```diff
@@ -0,0 +1,48 @@
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-p", "--port", default=8060, type=int)
+    parser.add_argument("-r", "--root", default="../../bin", type=str)
```
```suggestion
    import pathlib # somewhere on top
    parser.add_argument("-r", "--root", default="../../bin", type=pathlib.Path)
```

yedpodtrzitko commented on platform/javascript/serve.py:
```diff
@@ -0,0 +1,48 @@
+    browser_parser = parser.add_mutually_exclusive_group(required=False)
+    browser_parser.add_argument('-n', '--no-browser', dest='browser', action='store_false')
+    parser.set_defaults(browser=True)
+    parser.set_defaults(browser=True)
```
this line is here twice

yedpodtrzitko commented on platform/javascript/serve.py:
```diff
@@ -0,0 +1,48 @@
+if __name__ == "__main__":
+    # Change to the directory where the script is located,
+    # so that the script can be run from any location.
+    os.chdir(os.path.dirname(os.path.realpath(__file__)))
```
how about keeping this close to `if args.root: os.chdir(args.root)` lower, since it's related together. Maybe it can be there in `else:`?

yedpodtrzitko replied:
```suggestion
    os.chdir(Path(__file__).resolve().parent)
```

together with `from pathlib import Path`

yedpodtrzitko commented on platform/javascript/serve.py:
```diff
@@ -0,0 +1,48 @@
+class CORSRequestHandler(SimpleHTTPRequestHandler):
+    def end_headers(self):
+        self.send_header("Cross-Origin-Opener-Policy", "same-origin")
+        self.send_header("Cross-Origin-Embedder-Policy", "require-corp")
```
previous implementation was sending `Access-Control-Allow-Origin: *` too

Calinou commented:
I applied the suggestions and added help messages for command line options (visible with `--help` or `-h`).

Faless commented:
@Calinou you seem to have deleted the whole `package.json` which is used for linting and docs.
Only the `serve` dependency should be removed (and the `package.lock` updated accordingly).

Calinou commented:
> @Calinou you seem to have deleted the whole `package.json` which is used for linting and docs. Only the `serve` dependency should be removed (and the `package.lock` updated accordingly).

Fixed. I also renamed the `.gitignore` path so it ignores `node_modules/` correctly.

dzil123 commented on platform/web/serve.py:
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python3
+
+from http.server import HTTPServer, SimpleHTTPRequestHandler, test
```
```suggestion
from http.server import HTTPServer, SimpleHTTPRequestHandler, test  # type: ignore
```

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.6.
